thunderbirdparser/src/org/sleuthkit/autopsy/thunderbirdparser/EmailMessageThreader.java

```diff
@@ -0,0 +1,674 @@
+                subject_table.put(rootSubject, rootSetContainer);
+                finalSet.add(rootSetContainer);
+                continue;
+            }
+
+            // rootSetContainer and tableContainer either both have 'RE' or
+            // don't.  Create a new dummy container with both containers as
+            // children.
+            Container newParent = new Container();
+            newParent.addChild(tableContainer);
+            newParent.addChild(rootSetContainer);
+            subject_table.remove(rootSubject, tableContainer);
+            subject_table.put(rootSubject, newParent);
+
+            finalSet.add(newParent);
+        }
+        return finalSet;
+    }
+
+    /**
+     * Creates a Hashtable of Container and subjects. There will be one
+     * Container subject pair for each unique subject.
+     *
+     * @param rootSet The set of "root" Containers
+     *
+     * @return
+     */
+    HashMap<String, Container> createSubjectTable(Set<Container> rootSet) {
+        HashMap<String, Container> subject_table = new HashMap<>();
+
+        for (Container rootSetContainer : rootSet) {
+            String subject = "";
+            boolean reSubject = false;
+
+            if (rootSetContainer.hasMessage()) {
+                subject = rootSetContainer.getMessage().getSimplifiedSubject();
+                reSubject = rootSetContainer.getMessage().isReplySubject();
+            } else if (rootSetContainer.hasChildren()) {
+                Iterator<Container> childrenIterator = rootSetContainer.getChildren().iterator();
+                while (childrenIterator.hasNext()) {
+                    EmailMessage childMessage = childrenIterator.next().getMessage();
+                    if (childMessage != null) {
+                        subject = childMessage.getSimplifiedSubject();
+                        if (!subject.isEmpty()) {
+                            reSubject = childMessage.isReplySubject();
+                            break;
+                        }
+                    }
+                }
+            }
+
+            if (subject == null || subject.isEmpty()) {
+                continue;  // Give up on this container
+            }
+
+            Container tableContainer = subject_table.get(subject);
+
+            if (tableContainer == null
+                    || // Not in table
+                    (tableContainer.getMessage() != null && rootSetContainer.getMessage() == null)
+                    || // One in the table is empty, but current is not
+                    (!reSubject && (tableContainer.getMessage() != null && tableContainer.getMessage().isReplySubject()))) { //current doesn't have RE in it, use current instead of the one in the table
+                subject_table.put(subject, rootSetContainer);
+            }
+
+        }
+
+        return subject_table;
+    }
+    
+    private void assignThreadIDs(Set<Container> containerSet, String IDPrefix) {
```

Comment:
Add JavaDoc. 

is 'containerSet' at this point the same thing that we've been calling a 'rootSet'?

Reply:
Yes.  I believe so.